// zodiac-roles-sdk/build/cjs/evm/typechain-types/factories/contracts/adapters/Types.sol/ITransactionUnwrapper__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ITransactionUnwrapper__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "to",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "value",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
            {
                internalType: "enum Enum.Operation",
                name: "operation",
                type: "uint8",
            },
        ],
        name: "unwrap",
        outputs: [
            {
                components: [
                    {
                        internalType: "enum Enum.Operation",
                        name: "operation",
                        type: "uint8",
                    },
                    {
                        internalType: "address",
                        name: "to",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "dataLocation",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "dataSize",
                        type: "uint256",
                    },
                ],
                internalType: "struct UnwrappedTransaction[]",
                name: "result",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class ITransactionUnwrapper__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.ITransactionUnwrapper__factory = ITransactionUnwrapper__factory;
ITransactionUnwrapper__factory.abi = _abi;
