// zodiac-roles-sdk/build/esm/evm/typechain-types/factories/contracts/adapters/Types.sol/ITransactionUnwrapper__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, Interface } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "to",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "value",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
            {
                internalType: "enum Enum.Operation",
                name: "operation",
                type: "uint8",
            },
        ],
        name: "unwrap",
        outputs: [
            {
                components: [
                    {
                        internalType: "enum Enum.Operation",
                        name: "operation",
                        type: "uint8",
                    },
                    {
                        internalType: "address",
                        name: "to",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "dataLocation",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "dataSize",
                        type: "uint256",
                    },
                ],
                internalType: "struct UnwrappedTransaction[]",
                name: "result",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class ITransactionUnwrapper__factory {
    static abi = _abi;
    static createInterface() {
        return new Interface(_abi);
    }
    static connect(address, runner) {
        return new Contract(address, _abi, runner);
    }
}
