// zodiac-roles-sdk/build/cjs/sdk/src/permissions/authoring/conditions/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.describeStructure = exports.parameterType = exports.encodeValue = void 0;
const zodiac_roles_deployments_1 = require("zodiac-roles-deployments");
const encodeAbiParameters_1 = require("../../../utils/encodeAbiParameters");
const encodeValue = (value, type) => {
    return (0, encodeAbiParameters_1.encodeAbiParameters)([type], [value]);
};
exports.encodeValue = encodeValue;
const parameterType = (type) => {
    switch (type.baseType) {
        case "tuple":
            return zodiac_roles_deployments_1.ParameterType.Tuple;
        case "array":
            return zodiac_roles_deployments_1.ParameterType.Array;
        case "string":
        case "bytes":
            return zodiac_roles_deployments_1.ParameterType.Dynamic;
        default:
            return zodiac_roles_deployments_1.ParameterType.Static;
    }
};
exports.parameterType = parameterType;
const describeStructure = (type) => {
    const children = type.arrayChildren
        ? [type.arrayChildren]
        : type.components; // ethers typings are wrong
    return {
        paramType: (0, exports.parameterType)(type),
        operator: zodiac_roles_deployments_1.Operator.Pass,
        children: children && children.length > 0
            ? children.map(exports.describeStructure)
            : undefined,
    };
};
exports.describeStructure = describeStructure;
