// @gnosis-guild/eth-sdk/dist/abi-management/etherscan/explorerEndpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.predefinedExplorerEndpoints = void 0;
const networks_1 = require("../networks");
const makeAbiPubUrl = (networkId) => `https://api.abi.pub/v1/chains/${networkId}/etherscan`;
exports.predefinedExplorerEndpoints = {
    mainnet: {
        url: makeAbiPubUrl(networks_1.NetworkID.MAINNET),
        apiKey: '',
    },
    holesky: {
        url: makeAbiPubUrl(networks_1.NetworkID.HOLESKY),
        apiKey: '',
    },
    sepolia: {
        url: makeAbiPubUrl(networks_1.NetworkID.SEPOLIA),
        apiKey: '',
    },
    gnosis: {
        url: makeAbiPubUrl(networks_1.NetworkID.GNOSIS),
        apiKey: '',
    },
    optimism: {
        url: makeAbiPubUrl(networks_1.NetworkID.OPTIMISTIC_ETHEREUM),
        apiKey: '',
    },
    arbitrumOne: {
        url: makeAbiPubUrl(networks_1.NetworkID.ARBITRUM_ONE),
        apiKey: '',
    },
    base: {
        url: makeAbiPubUrl(networks_1.NetworkID.BASE),
        apiKey: '',
    },
    bsc: {
        url: makeAbiPubUrl(networks_1.NetworkID.BSC),
        apiKey: '',
    },
    polygon: {
        url: makeAbiPubUrl(networks_1.NetworkID.POLYGON),
        apiKey: '',
    },
    avalanche: {
        url: makeAbiPubUrl(networks_1.NetworkID.AVALANCHE),
        apiKey: '',
    },
    celo: {
        url: makeAbiPubUrl(networks_1.NetworkID.CELO),
        apiKey: '',
    },
};
