// zodiac-roles-sdk/build/cjs/evm/typechain-types/factories/contracts/Decoder__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decoder__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "CalldataOutOfBounds",
        type: "error",
    },
];
const _bytecode = "0x60556032600b8282823980515f1a607314602657634e487b7160e01b5f525f60045260245ffd5b305f52607381538281f3fe730000000000000000000000000000000000000000301460806040525f80fdfea2646970667358221220894feeaa2b169730e1338b28f089956958f379612fbd0bdc206dea97527a883364736f6c63430008150033";
const isSuperArgs = (xs) => xs.length > 1;
class Decoder__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.Decoder__factory = Decoder__factory;
Decoder__factory.bytecode = _bytecode;
Decoder__factory.abi = _abi;
